field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -0,0 +1,382 @@
+integer constants   I
+constants           c       ::= B | F | V
+universes           U       ::= Bool | Field | Vector
+statements          S, T    ::= let x = M | assert M
+                              | for var = I to I do S* end
+expressions         L, M, N ::= c | x | L binOp M | unOp L | L [ M ] | S ; L
+                              | if L then M else N
+binary operations   binOp   ::= or | ...
+unary operations    unOp    ::= not | ...
+@
+
+We assume the followin metafunctions:
+
+* tyConst(c), which returns the type of a constant,
+
+* tyVar(x), which returns the type of a variable,
+
+* tyLeft(binOp), which returns the type of the first argument of binOp,
+
+* tyRight(binOp), which returns the type of the second argument of binOp,
+
+* tyResult(binOp), which returns the type of the result of binOp,
+
+* tyArg(unOp), which returns the type of the argument of unOp,
+
+* tyResult(unOp), which returns the type of the result of unOp.
+
+We split the type-checking into the following judgements:
+
+* Type checking judgement which checks the type of expression, for
+  which both the expression @M@ and the universe @U@ are inputs:
+  @ M^+ <= U^+ @
+
+* Type synthesis (inference) judgement which infers the type of the
+  expression, for which the expression @M@ is an input and the
+  universe @U@ is an output:
+  @ M^+ => U^- @
+
+* Type equality judgement which checks wether two universes match, for
+  which both the universe @U1@ and @U2@ are inputs:
+  @ U1^+ = U2^+ @
+
+We can switch from type checking judgement to type inference judgement
+using type equality
+
+@
+M => U2
+U2 = U1
+-------
+M <= U1
+@
+
+Please note that at the moment we lack type ascriptions and for
+convenience we add type inference rule for if-then-else expressions.
+
+-}
+
+mkSomeUniVar :: forall f. T.Var -> T.SomeUniVar f
+mkSomeUniVar var
+    | '?':_ <- _varName var = Some $ T.UniVar Bool   var
+    | '#':_ <- _varName var = Some $ T.UniVar Vector var
+    | otherwise             = Some $ T.UniVar Field  var
+
+{-| Type inference for variables
+-}
+inferUniVar
+    :: forall m f. (MonadSupply m, MonadScope m) => R.Var -> m (T.SomeUniVar f)
+inferUniVar = liftM mkSomeUniVar . makeVar . R.unVar
+
+{-| Type inference for expressions
+@
+M => U
+@
+-}
+inferExpr
+    :: forall m f. (MonadTypeChecker m, TextField f)
+    => R.Expr R.Var f -> m (T.SomeUniExpr f)
+{-|
+@
+---------------
+c => tyConst(c)
+@
+-}
+inferExpr (R.EConst (Some c@(T.UniConst uni _))) =
+    pure $ SomeOf uni $ T.EConst c
+{-|
+@
+-------------
+x => tyVar(x)
+@
+-}    
+inferExpr (R.EVar   v) = do
+    Some uniVar@(T.UniVar uni _) <- inferUniVar v
+    pure $ SomeOf uni $ T.EVar uniVar
+{-|
+@
+L <= tyLeft(binOp)
+M <= tyRight(binOp)
+----------------------------
+L binOp M => tyResult(binOp)
+@
+-}    
+inferExpr (R.EAppBinOp rBinOp l m) =
+    withTypedBinOp rBinOp $
+    \tBinOp ->
+        SomeOf knownUni <$> (T.EAppBinOp tBinOp <$> checkExpr l <*> checkExpr m)
```

Comment:
Terrible nitpicking, but I find the lambda moved to the next line without any additional indentation a little unpleasant. I'd prefer `withTypedBinOp rBinOp $ \tBinOp ->`, i.e. keep it on the same line. Or move it to the next one, but indent it then (and then additionally indent the last line as well).

But feel free to ignore this comment, it's just nitpicking.